keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,425 @@
+
+The reservation of a pod is the union of all resource calculations of this form. Resources are considered individually when making decisions about admission or rejection from nodes or quota, and any one resource being exceeded will prevent execution or admission.
+
+For instance, the pod:
+
+```yaml
+metadata:
+  name: nginx
+spec:
+  initContainers:
+  - name: copy-files
+    ...
+    resources:
+      requests:
+        cpu: 100m
+  - name: generate-config
+    ...
+    resources:
+      requests:
+        cpu: 300m
+  containers:
+  - name: proxy
+    ...
+    resources:
+      requests:
+        cpu: 50m
+  - name: sidecar-logging
+    ...
+    resources:
+      requests:
+        cpu: 100m
+```
+
+will be handled by the system in the following ways:
+
+* Is considered to consume `300m` cores of CPU by the scheduler (the node must be able to run the largest init container by itself, which is larger than the sum of the proxy and sidecar containers at runtime)
+* The kubelet will not admit the pod if `(the node's allocatable CPU) - (the sum of all running pods)` is less than `300m`
+* The kubelet will create a cgroup for the pod that expects to get roughly `300m` cores, but the container cgroup created for `copy-files` will request `100m` while that init container is running
+* The quota subsystem will block this pod from being created if there is less than `300m` available.
+
+Once a pod is created, it passes through three high level lifecycle phases as seen by the total Kubernetes system. The first phase is `Pending` - the time before the pod is scheduled to a node (this is denoted by the nodeName field being set, not by the status update made by the Kubelet to the phase). The second phase is `Running`, when the Kubelet can begin initializing resources on the node and may start or restart process, and continues until the pod is deleted or reaches a terminal state of success or failure.  The final state is `Completed` which means the pod has no running containers, all resources have been released or cleaned up, and the pod will never again consume those resources. A pod can be considered in the `Completed` terminal phase when pod status phase has reached Succeeded or Failed, or the pod being is marked for deletion and all containers are reported stopped via status. We define one sub phase of `Running` - the `Initializing` phase is the time between when the pod is scheduled on a node and when all init containers have completed successfully and the `Initialized` condition on the pod is set to `true`. A pod without init containers may only remain in this phase very briefly. Because the Kubelet is not required to write pod status before it begins execution of a container, the `Initializing` phase is not consider a distinct top level phase (it is not distiguishable safely from a system perspective) except when presented by the Kubelet directly via metrics or on a node.
```

Comment:
Actually, where is this definition of Pending coming from?

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase differs

Reply:
Yes, this KEP is clarifying the general definition, and the doc you mentioned is wrong (it doesn't actually mean that) and will have to be corrected.  Pod phase is the reported value, but to the system the pod enters phases before they are reported, which means you can't make certain assumptions.  For example:

> The Pod has been accepted by the Kubernetes cluster, but one or more of the containers has not been set up and made ready to run. This includes time a Pod spends waiting to be scheduled as well as the time spent downloading container images over the network.

is inaccurate because there is no guarantee pod.status is ever set (ever).  The correct definition

> The Pod has been accepted by the Kubernetes cluster, but one or more of the containers has not yet been reported as running. This includes time a Pod spends waiting to be scheduled as well as the time spent downloading container images over the network.

and then a more general comment along the lines of:

> Pod phase is reported by the Kubelet and may be arbitrarily delayed by network issues or other temporary faults. Clients must consider pods that are scheduled to a node as possibly running except in the following states: when a pod is Succeeded or Failed, or when a pod has been marked deleted and all containers are reported as terminated.  Then the pod can be considered to be completed.

would probably get added to the API definition.

Reply:
The new definition that you are still proposing in the comment above (which I totally agree with) still differs from what the KEP is saying.

Reply:
Can you highlight what you see as different?  After the updates I did not expect it to.

Reply:
In the KEP:

```
The first phase is `Pending` - the time before the pod is scheduled to a node
```

In your comment above:

```
This includes time a Pod spends waiting to be scheduled as well as the time spent downloading container images over the network.
```

Reply:
Ah, I should clarify in the second comment that this is about the pod status phase.

> The Pod has been accepted by the Kubernetes cluster, but one or more of the containers has not yet been reported as running. This includes time a Pod spends waiting to be scheduled, time spent downloading container images over the network, and any delays or latency the node experiences reporting this status.

Ultimately the phase seen in the API is not useful to all clients.  The doc you referenced needs to be updated to clarify that a client has to interpret the presence of scheduling as "pod *might* be running" and the presence of the status phase as "pod *has* been running at some point previously".  There have been several fairly serious bugs in clients over the years that assumed phase was atomic, especially in controllers.

Reply:
> I think the definition about what is Pending or Running is somewhat irrelevant in this KEP and should be removed to avoid confusion.

I disagree because of this discussion :)  Different sigs have different definitions and administrators can't decide is a problem that the metrics don't make sense without.

> Even though your definition of "running" is what we use for scheduling, it doesn't match the existing definition of the phase Running.  I'm not sure if we can just change that definition from an API perspective.

This is a system architecture definition and does not change the meaning of the API field, so it's certainly within our remit to clarify the terminology.  I chose to approach this from the system perspective which is that in the existing system, a user MUST assume pods are running regardless of phase once scheduled because of the distributed nature of the system.  So this is less a change than a "proof" - clarifying for all users that from API down there are only three phases (which is not a change, it's reality).  The clarification in the KEP can then be used to clarify language changes in other docs, but we don't need an API change to do that, the API change is justified within the existing system.

> I would just try to be more explicit and say "non-terminated scheduled Pods".

In broad use the reason for this change is that even that statement is wrong because we don't use that consistently, it's hard to understand, and it's not actually the right terminology (not-terminated is not consistently used unfortunately).  I'd be happy to add more clarification here especially around terminology, but I disagree that this is a better outcome.

@derekwaynecarr @liggitt do you want to weigh in on this?  Especially the distinction about following the chain of logic I'm advancing here on the definition of Running.

Reply:
Re reading the section, i think use of "phase" is confusing and i'm going to update with a change to "state", and then draw the explicit points you're concerned about here (the connection between external observer and phase)

Reply:
there are (at least two) types of audiences:
1. I want to wait until positive confirmation X is running (and err on the side of waiting too long)
2. I want to know as soon as X might be running (and err on the side of assuming it is too early)

I agree the docs about the sequence of spec.nodeName -> kubelet pod status -> succeeded/failed/deleted+container terminated status should be super explicit about the possible realities at each step. I’d like to see the overall pod lifecycle made accurate, and the implications framed in two ways:
* "when the API field(s) look like this, here are the realities that could exist"
* "if you want to be certain at least point X of pod lifecycle has been reached, wait for the API fields to look like A,B,C"

I think the only reason this is in this KEP is because this is wanting to be explicit about when a given pod’s requests/limits are included in the usage metrics. I agree use of "phase" is confusing, and would avoid shorthand or choose a term that is not an API field (e.g. "include requested resources from all containers that could possibly be consuming resources on a node").

Reply:
> I want to wait until positive confirmation X is running (and err on the side of waiting too long)

This is what we are supporting with phase "Running" today, even though it's not clear. But as Jordan is pointing out, this is different from requests/limits being reported in the usage metrics.

For me, the problem with the current KEP description is that it's, in a way, removing "downloading containers" from the Pending phase.

I agree that it's better to separate phases from the resource model.

Reply:
To address the use of `Running`.  Jordan feels it would be confusing to overload it over phase, i feel it's implicitly confusing to use another term like `Scheduled`.  Here's some questions a user might ask of our API:

Q: Is a pod using resources?
A: A pod consumes from the overall pool when it is created for the purposes of quota, and node resources when it is considered running or runnable, and consumes no resources when it is terminal.

Q: Is a pod using node resources?
A: A pod is using node resources when the system considers it running or runnable.

Q: Is a pod running?
A: Do you mean whether its processes are running (it could be running), or whether it has been seen to be running at some point (i've at least reached the point where a process started), or whether I can see how much resources it is using?

Q: Is a pod's processes running?
A: If you see it scheduled to a node, you must assume the processes MAY be running unless it is terminal

Q: Is a pod terminal?
A: You must look at three fields - deletionTimestamp + status.containers.terminated != nil, or status.phase = Succeeded or Failed

Other questions include human factors:

1. Does a user know that phase is a trailing indicator that can be arbitrarily delayed?
2. Does a user know that setting nodeName is the last indicator they might see before the pod is completed?
3. Does a user get more confused calling the state after scheduling something generic (Scheduled) or something relatable but a superset of the workload (Running)?
4. What is the most consistent terminology that clearly expresses intent to users?

An argument might be that we should have a more coherent status field that reflects state as defined here - I think that would be a follow on KEP (in fact, by clarifying the real model here, we open the door for a subsequent "how do we implement that"), but I think the clarification is absolutely required to have the right metrics.  Note that we can always amend ground truth naming (which has no reflection in an API) via a PR to this KEP as we proceed, so if we choose `Running` now and take feedback that `Scheduled` would be better (or a third word) over time, that at least gives us a record of discussion.

Reply:
Pushed an update that hits some of these points, and attempts to put phase in the proper box with a larger concept bracketing it.

Reply:
I still think using the same names that we have for phases is confusing, given that we are giving different definitions to them.

From my perspective, we have 4 states, instead of 3: Pending, Scheduled, Running, Terminated. The current phases definition merges Pending and Scheduled into a single phase. The definition in this KEP merges Scheduled and Running in a single State. They are both correct, in a way, because phases is trying to answer "is my container guaranteed to be running", whereas this KEP is answering "is my pod consuming quota".

For this KEP, we need a term for "Scheduled or Running", and that cannot be either of those words, otherwise the confusion would remain and propagate to our users when we document the metrics.

Reply:
> From my perspective, we have 4 states, instead of 3: Pending, Scheduled, Running, Terminated. ... For this KEP, we need a term for "Scheduled or Running", and that cannot be either of those words, otherwise the confusion would remain and propagate to our users when we document the metrics.

As to number of states, if we had a state field on pod status, we could report three states correctly.  Pods have a terminal state at which point they are irrelevant to scheduler completely and only relevant to kubelet for ownership and final cleanup of temporary data like logs.  I think to your point the transitions are what define the states, and scheduling and terminal are transition boundaries, so we can only have three states.  Running is an optional substate in your language (you may never observe it).  

I think `Scheduled` could be seen as an adequate term for the middle state (since Completed ends Scheduled, and Scheduled has no equivalent elsewhere). We could conceivably use `Unscheduled` or `Schedulable` as the first state, in which case you have no overlap of words.  

However, if we wanted to connect the middle state to the core scenario, we could use `Schedulable`, `Runnable`, and `Completed` which probably draws the connection I was looking for without having the naming overlap.  I could imagine an argument that `Schedulable` isn't always true (a custom scheduler may be waiting for other input), so there might be some variants of those words that work.  I like `Runnable` better than `Scheduled` for state 2, and you could also argue that some workloads might not start on a Kubelet until other conditions are met but that in practice "meeting the conditions to schedule" and "meeting the conditions to run" is still part of that state from the large system.

Reply:
`Runnable` works for me. @liggitt WDYT?

Side note: Please do open an issue in k/website or an entire new KEP to fix the phases.

Reply:
that seems like a good way to describe it to me. just to be clear, this is not proposing an API change, just settling on terminology for this feature, right?

Reply:
Terminology for this feature, then follow up clarification (without terms) for the API fields, then maybe a followup kep to discuss using that terminology across the whole project with terms.

Reply:
Updated wording (and clarified the running pod usage metric is about the true running phase as measured at the kubelet)